### DISCUSSION TITLE
save function in estimators.classification.TensorFlowV2Classifier

### DISCUSSION THREAD

**Discussion by rhea2801** [Trusted-AI/adversarial-robustness-toolbox#1231] [Ideas]
Issue: Running the "save" function on a TensorFlowV2Classifier fitted on data, throws a NotImplementedError each time, even after providing filename & path correctly.

Found the function defined twice in adversarial-robustness-toolbox/art/estimators/classification/tensorflow.py

**First definition on line 623:**
 
` 
def save(self, filename: str, path: Optional[str] = None) -> None: `

        """
        Save a model to file in the format specific to the backend framework. For TensorFlow, .ckpt is used.
        :param filename: Name of the file where to store the model.
        :param path: Path of the folder where to store the model. If no path is specified, the model will be stored in
                     the default data location of the library ART_DATA_PATH.
        """
        # pylint: disable=E0611
        from tensorflow.python import saved_model
        from tensorflow.python.saved_model import tag_constants
        from tensorflow.python.saved_model.signature_def_utils_impl import predict_signature_def `

        if path is None:
            full_path = os.path.join(config.ART_DATA_PATH, filename)
        else:
            full_path = os.path.join(path, filename)

        if os.path.exists(full_path):
            shutil.rmtree(full_path)

        builder = saved_model.builder.SavedModelBuilder(full_path)
        signature = predict_signature_def(
            inputs={"SavedInputPhD": self._input_ph},
            outputs={"SavedOutput": self._output},
        )
        builder.add_meta_graph_and_variables(
            sess=self._sess,
            tags=[tag_constants.SERVING],
            signature_def_map={"predict": signature},
        )
        builder.save()

        logger.info("Model saved in path: %s.", full_path)`

**This definition seems to be the correct one, but the function is later defined again on line 1369, as:**

`
       def save(self, filename: str, path: Optional[str] = None) -> None:`

        """
        Save a model to file in the format specific to the backend framework. For TensorFlow, .ckpt is used.
        :param filename: Name of the file where to store the model.
        :param path: Path of the folder where to store the model. If no path is specified, the model will be stored in
                     the default data location of the library ART_DATA_PATH.
        """
        raise NotImplementedError

**How to reproduce:** 

1. Create a regular classification model
2. Create the ART TensorflowV2 classifier using the above generated one (parameter model = model created in step 1)
3. Fit on training data
4. Call model_name.save( filename='filename', path='path')

**Comment by beat-buesser** [comment]
Hi @rhea2801 Thank you very much for using ART! This is the currently expected behavior because this method has not yet been implemented. Would you be interested to implement and contribute it to ART?

**Comment by rhea2801** [comment]
Sure, how do I go about doing that?

**Comment by beat-buesser** [comment]
@rhea2801 That would be great! The main steps would include:
- fork the repository
- go to branch `dev_1.8.0`
- implement the method `TensorFlowV2Classifier.save` similar to KerasClassifier.save` but adapted to TensorFlow v2 models.
- test you implementation
- open pull request

Let us know anytime if you have questions!
